examples/flutter_gallery/lib/demo/pan_and_zoom_demo_transform_interaction.dart

```diff
@@ -0,0 +1,460 @@
+      matrix = matrixTranslate(matrix, widget.initialTranslation);
+    }
+    if (widget.initialScale != null) {
+      matrix = matrixScale(matrix, widget.initialScale);
+    }
+    if (widget.initialRotation != null) {
+      matrix = matrixRotate(matrix, widget.initialRotation, Offset.zero);
+    }
+    return matrix;
+  }
+
+  // Return the scene point underneath the viewport point given.
+  static Offset fromViewport(Offset viewportPoint, Matrix4 transform) {
+    // On viewportPoint, perform the inverse transformation of the scene to get
+    // where the point would be in the scene before the transformation.
+    final Matrix4 inverseMatrix = Matrix4.inverted(transform);
+    final Vector3 untransformed = inverseMatrix.transform3(Vector3(
+      viewportPoint.dx,
+      viewportPoint.dy,
+      0,
+    ));
+    return Offset(untransformed.x, untransformed.y);
+  }
+
+  // Get the offset of the current widget from the global screen coordinates.
+  static Offset getOffset(BuildContext context) {
+    final RenderBox renderObject = context.findRenderObject();
+    return renderObject.localToGlobal(Offset.zero);
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    _visibleRect = widget.visibleRect != null
+      ? widget.visibleRect
+      : Rect.fromLTWH(0, 0, widget.size.width, widget.size.height);
+    _transform = _initialTransform;
+    _controller = AnimationController(
+      vsync: this,
+    );
+    _controllerReset = AnimationController(
+      vsync: this,
+    );
+  }
+
+  @override
+  Widget build (BuildContext context) {
+    if (widget.reset && _animationReset == null) {
+      _animationReset?.removeListener(_onAnimateReset);
+      _controllerReset.reset();
+      _animationReset = Matrix4Tween(
+        begin: _transform,
+        end: _initialTransform,
+      ).animate(_controllerReset);
+      _controllerReset.duration = const Duration(milliseconds: 400);
+      _animationReset.addListener(_onAnimateReset);
+      _controllerReset.forward();
+    }
+
+    // A GestureDetector allows the detection of panning and zooming gestures on
+    // its child, which is the CustomPaint.
+    return GestureDetector(
+      behavior: HitTestBehavior.opaque, // Necessary when translating off screen
+      onTapDown: widget.onTapDown == null ? null : (TapDownDetails details) => widget.onTapDown(fromViewport(details.globalPosition - getOffset(context), _transform)),
+      onTapUp: widget.onTapUp == null ? null : (TapUpDetails details) => widget.onTapUp(fromViewport(details.globalPosition - getOffset(context), _transform)),
+      onTap: widget.onTap,
+      onTapCancel: widget.onTapCancel,
+      onDoubleTap: widget.onDoubleTap,
+      onLongPress: widget.onLongPress,
+      onLongPressUp: widget.onLongPressUp,
+      onVerticalDragDown: widget.onVerticalDragDown == null ? null : (DragDownDetails details) => widget.onVerticalDragDown(fromViewport(details.globalPosition - getOffset(context), _transform)),
```

Comment:
It might be clearer to factor out a generic method that produces these closure values:

```
T _callbackValue<T>(T callback) {
  if (callback == null)
    return callback;
  return  (T details) => callback(fromViewport(details.globalPosition - getOffset(context), _transform));
}
```
And then:

```
onVerticalDragDown:  callbackValue<DragDownDetails>(widget.onVerticalDragDown),
```

This is assuming that you can define a local generic method. Otherwise you'd have to pass along more parameters.

Reply:
I can't seem to get this to work.  It won't let me write `details.globalPosition` when details is an unknown type.  Unfortunately there is no common parent class or interface for all of these different Details objects.

When I first wrote this, I was also trying things like having TransformInteraction inherit from GestureDetector, but I couldn't get anything like that working either.

Any thing else to try?

Reply:
Sorry, here's a version that should work:
```
ValueChanged<T> _callbackValue<T>(ValueChanged<T> callback) {
  if (callback == null)
    return callback;
  return (T details) => callback(fromViewport(details.globalPosition - getOffset(context), _transform));
}
```

Reply:
I need your help implementing this.  These callbacks have changed a bit so it needs to be slightly different again.  I'll show you later.